dali/pipeline/operators/arithmetic/arithmetic_meta.h

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_META_H_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_ARITHMETIC_META_H_
+
+#include <cstdint>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+#include "dali/core/cuda_utils.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/types.h"
+
+namespace dali {
+
+enum class ArithmeticOp : int {
+  plus,
+  minus,
+  add,
+  sub,
+  mul,
+  div,
+  mod,
+};
+
+DALI_HOST_DEV constexpr int GetOpArity(ArithmeticOp op) {
+  switch (op) {
+    case ArithmeticOp::plus:
+    case ArithmeticOp::minus:
+      return 1;
+    case ArithmeticOp::add:
+    case ArithmeticOp::sub:
+    case ArithmeticOp::mul:
+    case ArithmeticOp::div:
+    case ArithmeticOp::mod:
+      return 2;
+    default:
+      return -1;
+  }
+}
+
+/**
+ * @brief Type promotion rules
+ *
+ * @tparam L
+ * @tparam R
+ *
+ * The rules are:
+ * T      + T      = T
+ * floatX + T      = floatX           (where T is not a float)
+ * floatX + floatY = float(max(X, Y))
+ * intX   + intY   = int(max(X, Y))
+ * uintX  + uintY  = uint(max(X, Y))
+ * intX   + uintY  = int2Y            (if X <= Y)
+ * intX   + uintY  = intX             (if X > Y)
```

Comment:
I don't know if conditional type promotion is something that one could expect.

Reply:
This is exactly what numpy does, we just ditched the weird case of promoting operation on purely integer types to double 🤷‍♂